Documentation/Plugins.md

```diff
@@ -170,11 +170,15 @@ struct MyPlugin: BuildToolPlugin {
 
 The plugin script can import *Foundation* and other standard libraries, but in the current version of SwiftPM, it cannot import other libraries.
 
+##### Build Commands
+
 In this example, the returned command is of the type `buildCommand`, so it will be incorporated into the build system's command graph and will run if any of the output files are missing or if the contents of any of the input files have changed since the last time the command ran.
 
-Note that build tool plugins are always applied to a target, which is passed in the parameter to the entry point.  Only source module targets have source files, so a plugin that iterates over source files will commonly test that the target it was given conforms to `SourceModuleTarget`.
+The target to which the plugin applies is passed as the `target` parameter.  Only source module targets have source files, so a plugin that iterates over source files will commonly test that the target it was given conforms to `SourceModuleTarget`.
```

Comment:
```suggestion
Build tool plugins are always applied to a specific target. The target to which the plugin applies is passed as the `target` parameter.  Only source module targets have source files, so a plugin that iterates over source files will commonly test that the target it was given conforms to `SourceModuleTarget`.
```

Reply:
this is of course subjective, but I think it is clearer.

Reply:
IMO this additional sentence sets the context clearly to the rest of the paragraph.

Reply:
I understand why you're saying that.  However making that statement by itself raises the questions, “as opposed to what?  why are you telling me this? am I supposed to imagine what it might mean to have a build tool that doesn't apply to a target so I can rule out some misconception I may be harboring?”  It raises more questions than it answers, and the reader doesn't need this dangling fact in order to understand the phrase "The target to which the plugin applies."

**If** you're going to say this, it belongs in a section about _what a build tool is_, not in this section which is about _how to use a build tool_.

Reply:
sure. you are obviously welcome to come up with a different way to make this + the other sentence above a little less terse - that is my main feedback about both of them

Reply:
Then there are two issues we need to sort out:

1. I remember now why I removed the original wording, aside from the fact that they are extra: if you read them as though they must be adding information, they are actually wrong.  As phrased, it makes it sound as though each build tool plugin applies to one specific target.  In fact, a build tool plugin can “apply to” (a better phrase would be “be used to build”) any number of targets, including zero.   If you want me to say something about that, I can do it in the appropriate place (the section about what a build tool is), although I think any framing of “what a build tool is” is going to make that patently obvious.  The bigger problem is that it won't do anything to make *this sentence* less terse, which is what you say you want.
2. You and I seem to have utterly different value systems w.r.t. concision in documentation.  For me, concision is a virtue, provided no necessary information is omitted, out of order, or ambiguously presented.  As a user, every word I have to process and sort out to understand the meaning of documentation makes my job harder.  Therefore, I have no understanding of how to make the docs less terse without making them actively worse.  For me, and I'm not exaggerating, it would be better for me to add words like “@tomerd thought this was too terse so please insert some words here that don't change the meaning,” because anything else leads the reader to wonder what meaning they are supposed to extract from the additional words.  What I suggest, if you don't want to simply reject this PR, is that you accept some phrasing you consider to be accurate but “too terse” and then you make an additional change yourself so that you're comfortable with the style.

Reply:
happy to hear other opinions, I am not sure if I can add a new signal beyond my subjective opinion that I have already expressed

Reply:
And I respect that.  If you are unwilling to solve the problem by doing what I suggest in №2 above, then I need a specific suggestion that makes the sentence less terse and isn't actively wrong, because as I wrote, I don't understand how to satisfy your desire.

Reply:
generally, I would like to avoid rejecting PRs and instead work with the authors to fine tune the proposed changes such that everyone involved is satisfied.

perhaps a path forward could be to ask feedback from a few other folks, maybe they can help us find the right words that will make it less terse but still not over-wordy.

cc @abertelrud @TimTr @bitjammer

Reply:
cc @amartini51

Reply:
I tend to agree that the additional wording "_Build tool plugins are always applied to a specific target_" felt like it was trying to document things beyond this parameter, which is out of scope. Documenting "bigger" stuff than the current scope can get out of hand quickly, and could be done for almost any API or parameter. In this case, the use feels pretty self-explanatory to me.

Perhaps it is useful to add an adjective like "mandatory" in this case, as perhaps that solves the suggester's concern? Seems like there is a desire for emphasis more than additional, external documentation.

Reply:
@TimTr please note this is not API docs - this is a user guide

Reply:
@dabrahams I read the entire section again starting from "Implementing the build tool plugin script" and then the two "Build Commands" and "Prebuild Commands" sub-section. The latter seem to be focused on explaining the examples for build and pre-build while the paragraphs above give a more broad overview of how build tool plugins work. In the light, I would argue the whole paragraph "The target to which the plugin applies is passed as the target parameter. Only source module targets have source files, so a plugin that iterates over source files will commonly test that the target it was given conforms to SourceModuleTarget" should be hoisted to the top part (above  "Build Commands" sub-section) as it is not specific to the example. At that point we can decide if we want to give additional context such as "SwiftPM will provide context information when invoking the plugin, through the context and target parameters..."